### PR TITLE
Closes softlayer/sl-ember-components#209

### PR DESCRIPTION
**Pull request by juwara0**
Wrote tests for the sl-datepicker component, updated documentation and removed items mentioned in comments.

### CODE REVIEW

File: addon/components/sl-date-picker.js [hunk 1]
```diff
@@ -68,6 +68,13 @@ export default Ember.Component.extend( TooltipEnabled, {
     daysOfWeekDisabled: [],
 
     /**
+     * When true, the datepicker will not be displayed
```
Comment (notmessenger) [inline addon/components/sl-date-picker.js]: well....the input field will display but be disabled and not able to be interacted with in order to select the date selection calendar.  Perhaps tweak this working just a small bit?

File: addon/components/sl-date-picker.js [hunk 2]
```diff
@@ -166,6 +173,14 @@ export default Ember.Component.extend( TooltipEnabled, {
     orientation: 'auto',
 
     /**
+     * The placeholder text that the datepicker should show
+     * be disabled
```
Comment (notmessenger) [inline addon/components/sl-date-picker.js]: `be disabled`?

File: addon/components/sl-date-picker.js [hunk 3]
```diff
@@ -200,6 +215,13 @@ export default Ember.Component.extend( TooltipEnabled, {
     todayHighlight: false,
 
     /**
+     * The date either selected by the datepicker or entered by the user
+     *
+     * @type {?Date}
```
Comment (notmessenger) [inline addon/components/sl-date-picker.js]: Can this property also be a `String`?

**Comment by notmessenger** [comment]
What is the behavior of the underlying data picker component when either `setEndDate` or `setStartDate` are set via calls to `setEndDate()` and `setStartDate()`?  Does it "redraw" itself, removing a selection that may now be out of the range?  Need to ensure that our expectation that as observers these properties can change that the component utilizing can actually react to such changes (resetting of values) rather than they only being intended to be set once.

File: addon/templates/components/sl-date-picker.hbs [hunk 4]
```diff
@@ -1,11 +1,10 @@
 {{#if label}}
-    <label for={{sl-date-picker-view.elementId}}>{{label}}</label>
+    <label for={{inputElementId}}>{{label}}</label>
 {{/if}}
 
 {{input
     viewName="sl-date-picker-view"
```
Comment (notmessenger) [inline addon/templates/components/sl-date-picker.hbs]: If getting rid of this property in the `label` element than can get rid of its assignment here.

File: tests/dummy/app/templates/demos/sl-date-picker.hbs [hunk 5]
```diff
@@ -52,6 +48,10 @@
         Days of the week that should be disabled. Values are 0 (Sunday) to 6 (Saturday). Multiple values should be comma-separated.
     {{/property-text}}
 
+    {{#property-text name="disabled" type="string" default="false"}}
+        When true, the datepicker will not be displayed.
```
Comment (notmessenger) [inline tests/dummy/app/templates/demos/sl-date-picker.hbs]: Same comment about this property in the component - make sure to synchronize whatever the new property description becomes.

File: addon/components/sl-date-picker.js (hunk 2 above)
Comment (juwara0) [inline addon/components/sl-date-picker.js]: I have removed this. It was copy-pasted.

File: tests/dummy/app/templates/demos/sl-date-picker.hbs [hunk 6]
```diff
@@ -111,6 +115,10 @@
         When true, highlight the current date.
     {{/property-text}}
 
+    {{#property-text name="value" type="date"}}
```
Comment (notmessenger) [inline tests/dummy/app/templates/demos/sl-date-picker.hbs]: Same inquiry as to the property in the component - can this also be a `String`?

File: tests/unit/components/sl-date-picker-test.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
+import Ember from 'ember';
+import { moduleForComponent, test } from 'ember-qunit';
+
+moduleForComponent( 'sl-date-picker', 'Unit | Component | sl date picker', {
+    unit: true
+});
+
+test('it renders', function( assert ) {
```
Comment (notmessenger) [inline tests/unit/components/sl-date-picker-test.js]: https://github.com/softlayer/ember-style-guide/blob/master/javascript.md#block-statements

File: addon/components/sl-date-picker.js (hunk 1 above)
Comment (juwara0) [inline addon/components/sl-date-picker.js]: This wording has been updated.

File: tests/unit/components/sl-date-picker-test.js [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    // Creates the component instance
+    var component = this.subject();
+    assert.equal( component._state, 'preRender' );
+
+    // Renders the component to the page
+    this.render();
+    assert.equal( component._state, 'inDOM' );
+});
+
+test( 'Default class names are present', function( assert ) {
```
Comment (notmessenger) [inline tests/unit/components/sl-date-picker-test.js]: Either in this test or different one the classes applied to the ``{{input}}`` in the template need to be ensured as well.

**Comment by notmessenger** [comment]
The `label` property is not documented anywhere nor is it tested.

**Comment by notmessenger** [comment]
The `helpText` property is not documented anywhere nor is it tested.

**Comment by notmessenger** [comment]
Let's go ahead and test that the object returned by https://github.com/juwara0/sl-ember-components/blob/209/addon/components/sl-date-picker.js#L321 contains expected keys.

File: tests/dummy/app/templates/demos/sl-date-picker.hbs (hunk 5 above)
Comment (juwara0) [inline tests/dummy/app/templates/demos/sl-date-picker.hbs]: This wording has been updated.

File: tests/unit/components/sl-date-picker-test.js (hunk 7 above)
Comment (juwara0) [inline tests/unit/components/sl-date-picker-test.js]: spacing issue fixed

File: addon/components/sl-date-picker.js (hunk 3 above)
Comment (juwara0) [inline addon/components/sl-date-picker.js]: I changed this as it actually should be a string and not a Date.

File: tests/dummy/app/templates/demos/sl-date-picker.hbs (hunk 6 above)
Comment (juwara0) [inline tests/dummy/app/templates/demos/sl-date-picker.hbs]: I changed this as it actually should be a string and not a Date.

File: addon/templates/components/sl-date-picker.hbs (hunk 4 above)
Comment (juwara0) [inline addon/templates/components/sl-date-picker.hbs]: This assignment has been removed.

**Comment by juwara0** [comment]
Tests have been written, documentation has been updated and default properties have been assigned for  'label' and 'helpText'.

File: tests/unit/components/sl-date-picker-test.js (hunk 8 above)
Comment (juwara0) [inline tests/unit/components/sl-date-picker-test.js]: Tests for the classes applied to {{input}} have been added.

**Comment by juwara0** [comment]
A test has been added that verifies that the options object contains the expected keys.

**Comment by juwara0** [comment]
Refactored setStartDate() and setEndDate() into updateDateRange() and added tests to ensure that our expectation that as observers these properties can change and that the component utilizing can actually react to such changes (resetting of values) rather than they only being intended to be set once.

**Comment by notmessenger** [comment]
https://github.com/juwara0/sl-ember-components/blob/209/addon/components/sl-date-picker.js#L275 needs to be `const`

**Comment by juwara0** [comment]
https://github.com/juwara0/sl-ember-components/blob/209/addon/components/sl-date-picker.js#L275 has been changed to be const

**Pull request merged**